### DISCUSSION TITLE
SIWE Token Gating Display Message

### DISCUSSION THREAD

**Discussion by jasonandrewth** [rainbow-me/rainbowkit#2035] [General]
Hello,

I have succesfully implemented Siwe in Next-auth (like so: https://docs.login.xyz/integrations/nextauth.js?ref=blog.spruceid.com) and using alchemy I check for a Token Balance, if it's 0 the login does not resolve. Now using @rainbow-me/rainbowkit-siwe-next-auth package it just displays a generic 'couldnt be signed' Error message. how can I change this to a personalized Message that notifies a user the token hasn't been found? I generally like the UI and it would be nice to keep it in the modal.

I don't see any options to customize the Message or the Look of the sign message part, I feel like that's really important for @rainbow-me/rainbowkit-siwe-next-auth to actually be useful. Please let me know if there is any such option:)

**Comment by 0xmkh** [comment]
@jasonandrewth We currently don't have a way to pass custom error messages to the modal. Would it be possible for you to check user balance before opening the actual modal ? We kind of want to stay away from letting users passing custom error messages, since we have internalization (i18n) support to maintain as well.

**Reply by jasonandrewth** [reply]
Okay I see,  I do the check using next-auth in an API route. Is there any way to programmatically open / close the Modal depending on the API response? Using @rainbow-me/rainbowkit-siwe-next-auth everything is kind of already working out of the box (with the modal asking for signing the message opening automatically). is there any way to intercept that flow (where i could close and open a new modal for instance) while using the rainbowkit-siwe-next-auth plugin, or any way of customizing that experience? Otherwise I guess I would need to build my own client side login with wagmi. I guess my main question is if the @rainbow-me/rainbowkit-siwe-next-auth plugin is just limited to this rather simple flow of connect wallet -> sign message -> success / error 

Goal is sth like this functionality-wise: https://boredapeyachtclub.com/login

Here is my server code where I verify the signature:
`
    const siwe = new SiweMessage(JSON.parse(credentials?.message || "{}"));
    const nextAuthUrl = new URL(process.env.NEXTAUTH_URL as string);
    const result = await siwe.verify({
      signature: credentials?.signature || "",
      domain: nextAuthUrl.host,
      nonce: await getCsrfToken({ req: { headers: req.headers } }),
       })
        if (result.success) {

        console.log("success")

        const ownsNFT = await validateNFTOwnership2(siwe.address);
         if (!ownsNFT) {
    
           throw new Error('You do not have the required tokens');
                   
         } else  return {
           id: siwe.address,
          };
`

**Reply by 0xmkh** [reply]
You're totally right, the SIWE RainbowKit feature is only limited to connect wallet -> sign message -> success / error. In your case you'd have to create your own flow with SIWE, pretty similar to what Dune is doing [here](https://dune.com/auth/login). 

To answer whether you can close and open RainbowKit modal, we only support opening and not closing. Here is the docs for more reference https://www.rainbowkit.com/docs/modal-hooks.

We're looking to make our SIWE better in the future, but for now we're sticking to what we have as of now.

Let me know if this answers your question 👍
